exporters/otlp/lib/opentelemetry/exporters/otlp/exporter.rb

```diff
@@ -0,0 +1,252 @@
+              SUCCESS
+            when Net::HTTPServiceUnavailable, Net::HTTPTooManyRequests
+              response.body # Read and discard body
+              redo if backoff?(retry_after: response['Retry-After'], retry_count: retry_count += 1)
+              FAILURE
+            when Net::HTTPRequestTimeOut, Net::HTTPGatewayTimeOut, Net::HTTPBadGateway
+              response.body # Read and discard body
+              redo if backoff?(retry_count: retry_count += 1)
+              FAILURE
+            when Net::HTTPBadRequest, Net::HTTPClientError, Net::HTTPServerError
+              # TODO: decode the body as a google.rpc.Status Protobuf-encoded message when https://github.com/open-telemetry/opentelemetry-collector/issues/1357 is fixed.
+              response.body # Read and discard body
+              FAILURE
+            when Net::HTTPRedirection
+              @http.finish
+              handle_redirect(response['location'])
+              redo if backoff?(retry_after: 0, retry_count: retry_count += 1)
+            else
+              @http.finish
+              FAILURE
+            end
+          rescue Net::OpenTimeout, Net::ReadTimeout
+            retry if backoff?(retry_count: retry_count += 1)
+            return FAILURE
+          end
+        end
+
+        def handle_redirect(location)
+          # TODO: figure out destination and reinitialize @http and @path
+          location = response['location']
+          location = URI(location).relative? ? (@uri + location).to_s : location
+          initialize_http(location) # Assume this is our new destination from now on.
+        end
+
+        def untraced
+          @tracer.with_span(OpenTelemetry::Trace::Span.new) { yield }
+        end
+
+        def backoff?(retry_after: nil, retry_count:, reason:)
+          return false if retry_count > @max_retry_count
+
+          sleep_interval = nil
+          unless retry_after.nil?
+            sleep_interval =
+              begin
+                Integer(retry_after)
+              rescue ArgumentError
+                nil
+              end
+            sleep_interval ||=
+              begin
+                Time.httpdate(retry_after) - Time.now
+              rescue
+                nil
+              end
+            sleep_interval = nil unless sleep_interval&.positive?
+          end
+          sleep_interval ||= rand(2**retry_count)
+
+          sleep(sleep_interval)
+          true
+        end
+
+        def encode(span_data)
+          Opentelemetry::Proto::Collector::Trace::V1::ExportTraceServiceRequest.encode(
+            Opentelemetry::Proto::Collector::Trace::V1::ExportTraceServiceRequest.new(
+              resource_spans: [
+                Opentelemetry::Proto::Trace::V1::ResourceSpans.new(
+                  resource: Opentelemetry::Proto::Resource::V1::Resource.new(
+                    attributes: OpenTelemetry.tracer_provider.resource.label_enumerator.map { |key, value| as_otlp_key_value(key, value) },
```

Comment:
The spec makes concessions for having multiple tracer providers in a single process. It's possible for multiple tracer providers to share the same exporter, and in that case, resources can differ by span. We may have some other issues with spec compliance existing around this poorly understood usecase, but we could support it here by also grouping by `span.resource`.

Here's the relevant spec details:

https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#tracerprovider

> Normally, the TracerProvider is expected to be accessed from a central place. Thus, the API SHOULD provide a way to set/register and access a global default TracerProvider.

> Notwithstanding any global TracerProvider, some applications may want to or have to use multiple TracerProvider instances, e.g. to have different configuration (like SpanProcessors) for each (and consequently for the Tracers obtained from them), or because its easier with dependency injection frameworks. Thus, implementations of TracerProvider SHOULD allow creating an arbitrary number of TracerProvider instances.

Nothing that this is spec'd as *SHOULD* and not *MUST*.

This issue also has some relevant discussion: https://github.com/open-telemetry/opentelemetry-specification/issues/508.

Here's an example from JS: https://github.com/open-telemetry/opentelemetry-js/blob/master/packages/opentelemetry-exporter-collector/src/transform.ts#L263-L282

I think the multiple TracerProvider use case is better solved with Resource Scopes (https://github.com/open-telemetry/oteps/issues/78), but since it's currently part of the spec (as a SHOULD), I'm inclined to go with it for now. I'm willing to be talked out of this position though.

Reply:
Done